src/backend/gl/src/window/surfman.rs

```diff
@@ -0,0 +1,342 @@
+            .bind_surface_to_context(&mut context, surface)
+            .expect("TODO");
+
+        device.make_context_current(&context).expect("TODO");
+
+        // Create a surface with the given context
+        Surface {
+            renderbuffer: None,
+            swapchain: None,
+            context: Starc::new(RwLock::new(context)),
+            device: Starc::new(RwLock::new(device)),
+        }
+    }
+}
+
+impl hal::Instance<B> for Instance {
+    fn create(_: &str, _: u32) -> Result<Self, hal::UnsupportedBackend> {
+        Ok(Instance {
+            hardware_adapter: SM_CONN.with(|c| c.borrow().create_hardware_adapter().expect("TODO")),
+            low_power_adapter: SM_CONN
+                .with(|c| c.borrow().create_low_power_adapter().expect("TODO")),
+            software_adapter: SM_CONN.with(|c| c.borrow().create_software_adapter().expect("TODO")),
+        })
+    }
+
+    fn enumerate_adapters(&self) -> Vec<Adapter<B>> {
+        let mut adapters = Vec::with_capacity(3);
+
+        let context_attributes = Self::get_default_context_attributes();
+
+        for surfman_adapter in &[
+            &self.hardware_adapter,
+            // TODO: Enabling these causes a segfault for an unknown reason
+            // &self.low_power_adapter,
+            // &self.software_adapter,
+        ] {
+            // Create a surfman device
+            let mut device =
+                SM_CONN.with(|c| c.borrow().create_device(surfman_adapter).expect("TODO"));
+
+            // Create context descriptor
+            let context_descriptor = device
+                .create_context_descriptor(&context_attributes)
+                .expect("TODO");
+
+            // Create context
+            let context = device.create_context(&context_descriptor).expect("TODO");
+
+            // Wrap in Starc<RwLock<T>>
+            let context = Starc::new(RwLock::new(context));
+            let context_ = context.clone();
+            let device = Starc::new(RwLock::new(device));
+            let device_ = device.clone();
+
+            // Create gl container
+            let gl = GlContainer::from_fn_proc(
+                |symbol_name| {
+                    device_
+                        .write()
+                        .get_proc_address(&context_.read(), symbol_name)
+                        as *const _
+                },
+                device,
+                context,
+            );
+
+            // Create physical device
+            adapters.push(PhysicalDevice::new_adapter((), gl));
+        }
+
+        adapters
+    }
+
+    unsafe fn create_surface(
+        &self,
+        has_handle: &impl raw_window_handle::HasRawWindowHandle,
+    ) -> Result<Surface, window::InitError> {
+        Ok(self.create_surface_from_rwh(has_handle.raw_window_handle()))
+    }
+
+    unsafe fn destroy_surface(&self, surface: Surface) {
+        // Surface implments Drop and will clean up the surface when it gets dropped
+        drop(surface);
```

Comment:
I wonder if this is idiomatic, given that you can just declare it as `_surface: Surface`

Reply:
I thought about that, but it seems a little strange right off because it almost seems like the function doesn't do anything until you realize that `Surface` is owned and will therefore be dropped.

I'm good going either way with it, though. Wasn't really set on one way or the other.

Reply:
yeah, I think calling `drop()` explicitly has its own advantages. It's just weird to see, makes you suspicious.
The danger of doing `_surface` is that it may be tempting for someone who looks at the code and tries to refactor it, to remove the parameter if it's unused. This danger doesn't apply if we are implementing an existing trait (like we do here), so it's fine to do.